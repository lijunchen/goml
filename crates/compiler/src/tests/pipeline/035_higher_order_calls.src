fn double(x: int) -> int {
    x * 2
}

fn increment(x: int) -> int {
    x + 1
}

fn chooser(flag: bool) -> (int) -> int {
    if flag {
        double
    } else {
        increment
    }
}

fn main() {
    let chosen = chooser(true) in
    let applied = chosen(5) in
    let direct = chooser(false)(5) in
    let printer = string_println in
    let _ = printer(int_to_string(applied)) in
    let _ = printer(int_to_string(direct)) in
    ()
}
