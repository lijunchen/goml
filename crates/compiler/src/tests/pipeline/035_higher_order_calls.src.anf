fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn chooser(flag/2: bool) -> (int) -> int {
  if flag/2 {
    double
  } else {
    increment
  }
}

fn main() -> unit {
  let xs/3 = [double, increment] in
  let f/4 = array_get(xs/3, 0) in
  let g/5 = array_get(xs/3, 1) in
  let t5 = f/4(10) in
  let t4 = g/5(t5) in
  let t3 = int_to_string(t4) in
  let mtmp0 = string_println(t3) in
  let chosen/6 = chooser(true) in
  let applied/7 = chosen/6(5) in
  let t6 = chooser(false) in
  let direct/8 = t6(5) in
  let printer/9 = string_println in
  let t7 = int_to_string(applied/7) in
  let mtmp1 = printer/9(t7) in
  let t8 = int_to_string(direct/8) in
  let mtmp2 = printer/9(t8) in
  ()
}