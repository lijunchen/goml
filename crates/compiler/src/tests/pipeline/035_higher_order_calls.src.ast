fn double(x: int) -> int {
    x * 2
}

fn increment(x: int) -> int {
    x + 1
}

fn chooser(flag: bool) -> (int) -> int {
    if flag {
        double
    } else {
        increment
    }
}

fn main() {
    let xs = [double, increment] in
    let f = array_get(xs, 0) in
    let g = array_get(xs, 1) in
    let _ = string_println(int_to_string(g(f(10)))) in
    let chosen = chooser(true) in
    let applied = chosen(5) in
    let direct = chooser(false)(5) in
    let printer = string_println in
    let _ = printer(int_to_string(applied)) in
    let _ = printer(int_to_string(direct)) in
    ()
}

