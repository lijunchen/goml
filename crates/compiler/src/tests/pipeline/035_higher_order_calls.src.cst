FILE@0..601
  FN@0..40
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..9 "double"
    PARAM_LIST@9..18
      LParen@9..10 "("
      PARAM@10..16
        Lident@10..11 "x"
        Colon@11..12 ":"
        Whitespace@12..13 " "
        TYPE_INT@13..16
          IntKeyword@13..16 "int"
      RParen@16..17 ")"
      Whitespace@17..18 " "
    Arrow@18..20 "->"
    Whitespace@20..21 " "
    TYPE_INT@21..25
      IntKeyword@21..24 "int"
      Whitespace@24..25 " "
    BLOCK@25..40
      LBrace@25..26 "{"
      Whitespace@26..31 "\n    "
      EXPR_BINARY@31..37
        EXPR_LIDENT@31..33
          Lident@31..32 "x"
          Whitespace@32..33 " "
        Star@33..34 "*"
        Whitespace@34..35 " "
        EXPR_INT@35..37
          Int@35..36 "2"
          Whitespace@36..37 "\n"
      RBrace@37..38 "}"
      Whitespace@38..40 "\n\n"
  FN@40..83
    FnKeyword@40..42 "fn"
    Whitespace@42..43 " "
    Lident@43..52 "increment"
    PARAM_LIST@52..61
      LParen@52..53 "("
      PARAM@53..59
        Lident@53..54 "x"
        Colon@54..55 ":"
        Whitespace@55..56 " "
        TYPE_INT@56..59
          IntKeyword@56..59 "int"
      RParen@59..60 ")"
      Whitespace@60..61 " "
    Arrow@61..63 "->"
    Whitespace@63..64 " "
    TYPE_INT@64..68
      IntKeyword@64..67 "int"
      Whitespace@67..68 " "
    BLOCK@68..83
      LBrace@68..69 "{"
      Whitespace@69..74 "\n    "
      EXPR_BINARY@74..80
        EXPR_LIDENT@74..76
          Lident@74..75 "x"
          Whitespace@75..76 " "
        Plus@76..77 "+"
        Whitespace@77..78 " "
        EXPR_INT@78..80
          Int@78..79 "1"
          Whitespace@79..80 "\n"
      RBrace@80..81 "}"
      Whitespace@81..83 "\n\n"
  FN@83..193
    FnKeyword@83..85 "fn"
    Whitespace@85..86 " "
    Lident@86..93 "chooser"
    PARAM_LIST@93..106
      LParen@93..94 "("
      PARAM@94..104
        Lident@94..98 "flag"
        Colon@98..99 ":"
        Whitespace@99..100 " "
        TYPE_BOOL@100..104
          BoolKeyword@100..104 "bool"
      RParen@104..105 ")"
      Whitespace@105..106 " "
    Arrow@106..108 "->"
    Whitespace@108..109 " "
    TYPE_FUNC@109..122
      TYPE_TUPLE@109..115
        TYPE_LIST@109..115
          LParen@109..110 "("
          TYPE_INT@110..113
            IntKeyword@110..113 "int"
          RParen@113..114 ")"
          Whitespace@114..115 " "
      Arrow@115..117 "->"
      Whitespace@117..118 " "
      TYPE_INT@118..122
        IntKeyword@118..121 "int"
        Whitespace@121..122 " "
    BLOCK@122..193
      LBrace@122..123 "{"
      Whitespace@123..128 "\n    "
      EXPR_IF@128..190
        IfKeyword@128..130 "if"
        Whitespace@130..131 " "
        EXPR_IF_COND@131..136
          EXPR_LIDENT@131..136
            Lident@131..135 "flag"
            Whitespace@135..136 " "
        EXPR_IF_THEN@136..159
          BLOCK@136..159
            LBrace@136..137 "{"
            Whitespace@137..146 "\n        "
            EXPR_LIDENT@146..157
              Lident@146..152 "double"
              Whitespace@152..157 "\n    "
            RBrace@157..158 "}"
            Whitespace@158..159 " "
        ElseKeyword@159..163 "else"
        Whitespace@163..164 " "
        EXPR_IF_ELSE@164..190
          BLOCK@164..190
            LBrace@164..165 "{"
            Whitespace@165..174 "\n        "
            EXPR_LIDENT@174..188
              Lident@174..183 "increment"
              Whitespace@183..188 "\n    "
            RBrace@188..189 "}"
            Whitespace@189..190 "\n"
      RBrace@190..191 "}"
      Whitespace@191..193 "\n\n"
  FN@193..601
    FnKeyword@193..195 "fn"
    Whitespace@195..196 " "
    Lident@196..200 "main"
    PARAM_LIST@200..203
      LParen@200..201 "("
      RParen@201..202 ")"
      Whitespace@202..203 " "
    BLOCK@203..601
      LBrace@203..204 "{"
      Whitespace@204..209 "\n    "
      EXPR_LET@209..599
        LetKeyword@209..212 "let"
        Whitespace@212..213 " "
        PATTERN_VARIABLE@213..216
          Lident@213..215 "xs"
          Whitespace@215..216 " "
        Eq@216..217 "="
        Whitespace@217..218 " "
        EXPR_LET_VALUE@218..238
          EXPR_ARRAY_LITERAL@218..238
            LBracket@218..219 "["
            EXPR_LIDENT@219..225
              Lident@219..225 "double"
            Comma@225..226 ","
            Whitespace@226..227 " "
            EXPR_LIDENT@227..236
              Lident@227..236 "increment"
            RBracket@236..237 "]"
            Whitespace@237..238 " "
        InKeyword@238..240 "in"
        Whitespace@240..245 "\n    "
        EXPR_LET_BODY@245..599
          EXPR_LET@245..599
            LetKeyword@245..248 "let"
            Whitespace@248..249 " "
            PATTERN_VARIABLE@249..251
              Lident@249..250 "f"
              Whitespace@250..251 " "
            Eq@251..252 "="
            Whitespace@252..253 " "
            EXPR_LET_VALUE@253..270
              EXPR_CALL@253..270
                EXPR_LIDENT@253..262
                  Lident@253..262 "array_get"
                ARG_LIST@262..270
                  LParen@262..263 "("
                  ARG@263..267
                    EXPR_LIDENT@263..265
                      Lident@263..265 "xs"
                    Comma@265..266 ","
                    Whitespace@266..267 " "
                  ARG@267..268
                    EXPR_INT@267..268
                      Int@267..268 "0"
                  RParen@268..269 ")"
                  Whitespace@269..270 " "
            InKeyword@270..272 "in"
            Whitespace@272..277 "\n    "
            EXPR_LET_BODY@277..599
              EXPR_LET@277..599
                LetKeyword@277..280 "let"
                Whitespace@280..281 " "
                PATTERN_VARIABLE@281..283
                  Lident@281..282 "g"
                  Whitespace@282..283 " "
                Eq@283..284 "="
                Whitespace@284..285 " "
                EXPR_LET_VALUE@285..302
                  EXPR_CALL@285..302
                    EXPR_LIDENT@285..294
                      Lident@285..294 "array_get"
                    ARG_LIST@294..302
                      LParen@294..295 "("
                      ARG@295..299
                        EXPR_LIDENT@295..297
                          Lident@295..297 "xs"
                        Comma@297..298 ","
                        Whitespace@298..299 " "
                      ARG@299..300
                        EXPR_INT@299..300
                          Int@299..300 "1"
                      RParen@300..301 ")"
                      Whitespace@301..302 " "
                InKeyword@302..304 "in"
                Whitespace@304..309 "\n    "
                EXPR_LET_BODY@309..599
                  EXPR_LET@309..599
                    LetKeyword@309..312 "let"
                    Whitespace@312..313 " "
                    PATTERN_WILDCARD@313..315
                      WildcardKeyword@313..314 "_"
                      Whitespace@314..315 " "
                    Eq@315..316 "="
                    Whitespace@316..317 " "
                    EXPR_LET_VALUE@317..357
                      EXPR_CALL@317..357
                        EXPR_LIDENT@317..331
                          Lident@317..331 "string_println"
                        ARG_LIST@331..357
                          LParen@331..332 "("
                          ARG@332..355
                            EXPR_CALL@332..355
                              EXPR_LIDENT@332..345
                                Lident@332..345 "int_to_string"
                              ARG_LIST@345..355
                                LParen@345..346 "("
                                ARG@346..354
                                  EXPR_CALL@346..354
                                    EXPR_LIDENT@346..347
                                      Lident@346..347 "g"
                                    ARG_LIST@347..354
                                      LParen@347..348 "("
                                      ARG@348..353
                                        EXPR_CALL@348..353
                                          EXPR_LIDENT@348..349
                                            Lident@348..349 "f"
                                          ARG_LIST@349..353
                                            LParen@349..350 "("
                                            ARG@350..352
                                              EXPR_INT@350..352
                                                Int@350..352 "10"
                                            RParen@352..353 ")"
                                      RParen@353..354 ")"
                                RParen@354..355 ")"
                          RParen@355..356 ")"
                          Whitespace@356..357 " "
                    InKeyword@357..359 "in"
                    Whitespace@359..364 "\n    "
                    EXPR_LET_BODY@364..599
                      EXPR_LET@364..599
                        LetKeyword@364..367 "let"
                        Whitespace@367..368 " "
                        PATTERN_VARIABLE@368..375
                          Lident@368..374 "chosen"
                          Whitespace@374..375 " "
                        Eq@375..376 "="
                        Whitespace@376..377 " "
                        EXPR_LET_VALUE@377..391
                          EXPR_CALL@377..391
                            EXPR_LIDENT@377..384
                              Lident@377..384 "chooser"
                            ARG_LIST@384..391
                              LParen@384..385 "("
                              ARG@385..389
                                EXPR_BOOL@385..389
                                  TrueKeyword@385..389 "true"
                              RParen@389..390 ")"
                              Whitespace@390..391 " "
                        InKeyword@391..393 "in"
                        Whitespace@393..398 "\n    "
                        EXPR_LET_BODY@398..599
                          EXPR_LET@398..599
                            LetKeyword@398..401 "let"
                            Whitespace@401..402 " "
                            PATTERN_VARIABLE@402..410
                              Lident@402..409 "applied"
                              Whitespace@409..410 " "
                            Eq@410..411 "="
                            Whitespace@411..412 " "
                            EXPR_LET_VALUE@412..422
                              EXPR_CALL@412..422
                                EXPR_LIDENT@412..418
                                  Lident@412..418 "chosen"
                                ARG_LIST@418..422
                                  LParen@418..419 "("
                                  ARG@419..420
                                    EXPR_INT@419..420
                                      Int@419..420 "5"
                                  RParen@420..421 ")"
                                  Whitespace@421..422 " "
                            InKeyword@422..424 "in"
                            Whitespace@424..429 "\n    "
                            EXPR_LET_BODY@429..599
                              EXPR_LET@429..599
                                LetKeyword@429..432 "let"
                                Whitespace@432..433 " "
                                PATTERN_VARIABLE@433..440
                                  Lident@433..439 "direct"
                                  Whitespace@439..440 " "
                                Eq@440..441 "="
                                Whitespace@441..442 " "
                                EXPR_LET_VALUE@442..460
                                  EXPR_CALL@442..460
                                    EXPR_CALL@442..456
                                      EXPR_LIDENT@442..449
                                        Lident@442..449 "chooser"
                                      ARG_LIST@449..456
                                        LParen@449..450 "("
                                        ARG@450..455
                                          EXPR_BOOL@450..455
                                            FalseKeyword@450..455 "false"
                                        RParen@455..456 ")"
                                    ARG_LIST@456..460
                                      LParen@456..457 "("
                                      ARG@457..458
                                        EXPR_INT@457..458
                                          Int@457..458 "5"
                                      RParen@458..459 ")"
                                      Whitespace@459..460 " "
                                InKeyword@460..462 "in"
                                Whitespace@462..467 "\n    "
                                EXPR_LET_BODY@467..599
                                  EXPR_LET@467..599
                                    LetKeyword@467..470 "let"
                                    Whitespace@470..471 " "
                                    PATTERN_VARIABLE@471..479
                                      Lident@471..478 "printer"
                                      Whitespace@478..479 " "
                                    Eq@479..480 "="
                                    Whitespace@480..481 " "
                                    EXPR_LET_VALUE@481..496
                                      EXPR_LIDENT@481..496
                                        Lident@481..495 "string_println"
                                        Whitespace@495..496 " "
                                    InKeyword@496..498 "in"
                                    Whitespace@498..503 "\n    "
                                    EXPR_LET_BODY@503..599
                                      EXPR_LET@503..599
                                        LetKeyword@503..506 "let"
                                        Whitespace@506..507 " "
                                        PATTERN_WILDCARD@507..509
                                          WildcardKeyword@507..508 "_"
                                          Whitespace@508..509 " "
                                        Eq@509..510 "="
                                        Whitespace@510..511 " "
                                        EXPR_LET_VALUE@511..543
                                          EXPR_CALL@511..543
                                            EXPR_LIDENT@511..518
                                              Lident@511..518 "printer"
                                            ARG_LIST@518..543
                                              LParen@518..519 "("
                                              ARG@519..541
                                                EXPR_CALL@519..541
                                                  EXPR_LIDENT@519..532
                                                    Lident@519..532 "int_to_string"
                                                  ARG_LIST@532..541
                                                    LParen@532..533 "("
                                                    ARG@533..540
                                                      EXPR_LIDENT@533..540
                                                        Lident@533..540 "applied"
                                                    RParen@540..541 ")"
                                              RParen@541..542 ")"
                                              Whitespace@542..543 " "
                                        InKeyword@543..545 "in"
                                        Whitespace@545..550 "\n    "
                                        EXPR_LET_BODY@550..599
                                          EXPR_LET@550..599
                                            LetKeyword@550..553 "let"
                                            Whitespace@553..554 " "
                                            PATTERN_WILDCARD@554..556
                                              WildcardKeyword@554..555 "_"
                                              Whitespace@555..556 " "
                                            Eq@556..557 "="
                                            Whitespace@557..558 " "
                                            EXPR_LET_VALUE@558..589
                                              EXPR_CALL@558..589
                                                EXPR_LIDENT@558..565
                                                  Lident@558..565 "printer"
                                                ARG_LIST@565..589
                                                  LParen@565..566 "("
                                                  ARG@566..587
                                                    EXPR_CALL@566..587
                                                      EXPR_LIDENT@566..579
                                                        Lident@566..579 "int_to_string"
                                                      ARG_LIST@579..587
                                                        LParen@579..580 "("
                                                        ARG@580..586
                                                          EXPR_LIDENT@580..586
                                                            Lident@580..586 "direct"
                                                        RParen@586..587 ")"
                                                  RParen@587..588 ")"
                                                  Whitespace@588..589 " "
                                            InKeyword@589..591 "in"
                                            Whitespace@591..596 "\n    "
                                            EXPR_LET_BODY@596..599
                                              EXPR_UNIT@596..599
                                                LParen@596..597 "("
                                                RParen@597..598 ")"
                                                Whitespace@598..599 "\n"
      RBrace@599..600 "}"
      Whitespace@600..601 "\n"