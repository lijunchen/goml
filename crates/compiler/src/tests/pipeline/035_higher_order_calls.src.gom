package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_mul(x int, y int) int {
    return x * y
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func array_get__Array_2_Fn_int_to_int(arr [2]func(int) int, index int) func(int) int {
    return arr[index]
}

func double(x__0 int) int {
    var ret9 int
    ret9 = int_mul(x__0, 2)
    return ret9
}

func increment(x__1 int) int {
    var ret10 int
    ret10 = int_add(x__1, 1)
    return ret10
}

func chooser(flag__2 bool) func(int) int {
    var ret11 func(int) int
    if flag__2 {
        ret11 = double
    } else {
        ret11 = increment
    }
    return ret11
}

func main0() struct{} {
    var ret12 struct{}
    var xs__3 [2]func(int) int = [2]func(int) int{double, increment}
    var f__4 func(int) int = array_get__Array_2_Fn_int_to_int(xs__3, 0)
    var g__5 func(int) int = array_get__Array_2_Fn_int_to_int(xs__3, 1)
    var t5 int = f__4(10)
    var t4 int = g__5(t5)
    var t3 string = int_to_string(t4)
    string_println(t3)
    var chosen__6 func(int) int = chooser(true)
    var applied__7 int = chosen__6(5)
    var t6 func(int) int = chooser(false)
    var direct__8 int = t6(5)
    var printer__9 func(string) struct{} = string_println
    var t7 string = int_to_string(applied__7)
    printer__9(t7)
    var t8 string = int_to_string(direct__8)
    printer__9(t8)
    ret12 = struct{}{}
    return ret12
}

func main() {
    main0()
}
