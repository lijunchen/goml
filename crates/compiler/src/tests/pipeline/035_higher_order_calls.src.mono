fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn chooser(flag/2: bool) -> (int) -> int {
  if flag/2 {
      double
  } else {
      increment
  }
}

fn main() -> unit {
  let xs/3 = [double, increment] in
  let f/4 = array_get(xs/3, 0) in
  let g/5 = array_get(xs/3, 1) in
  let mtmp0 = string_println(int_to_string(g/5(f/4(10)))) in
  let chosen/6 = chooser(true) in
  let applied/7 = chosen/6(5) in
  let direct/8 = chooser(false)(5) in
  let printer/9 = string_println in
  let mtmp1 = printer/9(int_to_string(applied/7)) in
  let mtmp2 = printer/9(int_to_string(direct/8)) in
  ()
}