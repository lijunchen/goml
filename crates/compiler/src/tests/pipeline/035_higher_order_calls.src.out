21
10
6
